### PR TITLE
Cache projectile-project-root for performance

### PR DESCRIPTION
**Pull request by alexmurray**
Cache the result of (projectile-project-root) as a buffer local variable to
increase performance - this is implemented the same as was done
for (projectile-project-name) in PR 1906.

Should help resolve issue #1003

### CODE REVIEW

**Comment by fommil** [comment]
I'd like to give this a try because I've been hit by perf in the past.

File: projectile.el [hunk 1]
```diff
@@ -860,29 +860,55 @@ topmost sequence of matched directories.  Nil otherwise."
...
-        (if projectile-require-project-root
-            (error "You're not in a project")
-          default-directory))))
+  (or (cl-subst nil 'none
+                (or (and (equal projectile-cached-buffer-file-name buffer-file-name)
+                         projectile-cached-project-root)
+                    (progn
+                      (setq projectile-cached-buffer-file-name buffer-file-name)
+                      (setq projectile-cached-project-root
+                            (or (let* ((dir default-directory)
```
Comment (bbatsov) [inline projectile.el]: You forgot the move the comment explaining this code here.

File: projectile.el [hunk 2]
```diff
@@ -911,10 +937,13 @@ This variable is reset automatically when Projectile detects that
 the `buffer-file-name' has changed. It can also be reset manually
 by calling `projectile-reset-cached-project-name'.")
 
-(defvar-local projectile-cached-buffer-file-name nil
```
Comment (bbatsov) [inline projectile.el]: Why did you delete this?

File: projectile.el [hunk 3]
```diff
@@ -860,29 +860,55 @@ topmost sequence of matched directories.  Nil otherwise."
...
-               (if (and cache-value (file-exists-p cache-value))
-                   cache-value
-                 (let ((value (funcall func (file-truename dir))))
-                   (puthash cache-key value projectile-project-root-cache)
-                   value))))
-           projectile-project-root-files-functions))
-        (if projectile-require-project-root
-            (error "You're not in a project")
-          default-directory))))
+  (or (cl-subst nil 'none
```
Comment (bbatsov) [inline projectile.el]: I don't understand what exactly is this supposed to do. Seems like some unnecessary complication of the code.

Comment (alexmurray) [inline projectile.el]: To handle the case where there is no project root (and (projectile-project-root) returns nil) we can't just store nil as the cached value - as then we would end up reevaluating it every time and never using the cache. So instead we store the symbol 'none there instead - however we don't want to return 'none from the function since this invalidates the expected contract (return nil if not in a project) so we use cl-subst to replace any 'none value with nil.

File: projectile.el (hunk 1 above)
Comment (alexmurray) [inline projectile.el]: Ah thanks.

File: projectile.el (hunk 2 above)
Comment (alexmurray) [inline projectile.el]: I just moved it up earlier in the file so it is declared before we now first reference it.

File: projectile.el (hunk 3 above)
Comment (bbatsov) [inline projectile.el]: Can't you simply use `'none` as the default value for the buffer-local-variable?

Comment (alexmurray) [inline projectile.el]: But then we will never evaluate it - the code is:

```
(or (and (equal projectile-cached-buffer-file-name buffer-file-name)
               projectile-cached-project-root)
```

So a non-nil value means the buffer-local cached value will be used - so it needs to be nil initially, then it gets evaluated once, and either set to a path or `'none` and then it can be used as a cached value without reevaluation in the future.

Comment (fommil) [inline projectile.el]: Isn't it easier if nil just cache misses? So what, projectile is slightly slower on directories that are not source managed or have a manual .projectile :-/

There is an elisp package for caching btw, if you'd like to reuse it.

Comment (alexmurray) [inline projectile.el]: I disagree - I often use emacs on directories which are not part of a project and having EVERYTHING slow down (say when using [spaceline-all-the-icons](https://github.com/alexmurray/spaceline-all-the-icons.el) which calls `(projectile-project-root)` on every redisplay) is very annoying. Why go to the effort of caching in a project and not also add support for caching outside of a project - also since when outside a project to determine this each time is expensive since we have to exhaust every possible project test each time - it MORE of a win to cache in this case than in the case of being in a project.

Which elisp package are you referring to?

Comment (bbatsov) [inline projectile.el]: Just explain in a comment what this code is doing for future reference and I'll merge it. No need for more 3rd party deps.

Comment (alexmurray) [inline projectile.el]: How about c473d17 ? Do you want me to squash these all into a single commit?

File: projectile.el (hunk 2 above)
Comment (bbatsov) [inline projectile.el]: No need for this. I can squash from the web ui.

**Comment by bbatsov** [comment]
Thanks!

File: projectile.el [hunk 4]
```diff
@@ -911,10 +945,13 @@ This variable is reset automatically when Projectile detects that
 the `buffer-file-name' has changed. It can also be reset manually
 by calling `projectile-reset-cached-project-name'.")
 
-(defvar-local projectile-cached-buffer-file-name nil
-  "The last known value of `buffer-file-name' for the current
-buffer. This is used to detect a change in `buffer-file-name',
-which triggers a reset of `projectile-cached-project-name'.")
+(defun projectile-reset-cached-project-root ()
```
Comment (aaronjensen) [inline projectile.el]: Would it make sense to call this during `projectile-invalidate-cache`?

**Comment by Silex** [comment]
I wonder if it still makes sense to cache `projectile-cached-project-name`

**Comment by fommil** [comment]
that lib: https://github.com/skeeto/emacs-memoize

**Comment by fommil** [comment]
(in hindsight, I agree it makes sense to cache non-projects. I have worked in environments with network drives and ... short of having an extensive exclusion list for projectile... a cache for buffers in those folders would have surely helped)

**Comment by JeanMax** [comment]
Hey!

This merge broke (projectile-all-project-files) for me.
It now only returns the files of the current project... so (projectile-find-file-in-known-projects) acts as (projectile-find-file); 'projectile-known-projects is set correctly.

I could reproduce the behaviour with 'emacs -q' and 2 random projects, but it seems that it doesn't affect everybody.
linux / emacs 25.2.1
